addons/payment_alipay/models/payment_transaction.py

```diff
@@ -0,0 +1,147 @@
+        if self.acquirer_id and self.acquirer_id.provider == 'alipay' and self.acquirer_id.alipay_payment_method == 'express_checkout':
+            if self.currency_id.name != 'CNY':
+                _logger.info("Only CNY currency is allowed for Alipay Express Checkout")
+                raise ValidationError("Alipay: " + _(
+                    "Only transactions in Chinese Yuan (CNY) are allowed for Express Checkout.\n"
+                    "If you wish to use another currency than CNY for your transactions, switch\n"
+                    "your configuration to a Cross-border account in Odoo."
+                ))
+
+    # === BUSINESS METHODS ===#
+
+    @api.model
+    def _get_tx_from_feedback_data(self, provider, data):
+        if provider != 'alipay':
+            return super()._get_tx_from_feedback_data(data, provider)
+        # verify signature
+        alipay_provider = self.env['payment.acquirer'].search([('provider', '=', 'alipay')], limit=1)
+        sign_check = alipay_provider._alipay_build_sign(data)
+        txn_id, sign = data.get('trade_no'), data.get('sign')
+        reference = data.get('reference') or data.get('out_trade_no')
+        if sign != sign_check:
+            _logger.info('Alipay: invalid sign, received %s, computed (%s), for data (%s)' % (sign, sign_check, data))
+            raise ValidationError('Alipay: ' + _('invalid sign: received %(sign)s, computed %(sc)s, for data %(data)s',
+                                                 sign=sign, sc=sign_check, data=data))
+
+        if not reference or not txn_id:
+            _logger.info('Alipay: received data with missing reference (%s) or txn_id (%s)' % (reference, txn_id))
+            raise ValidationError("Alipay: " +_('received data with missing reference %(r)s or txn_id %(t)s',
+                                                r=reference, t=txn_id))
+
+        tx = self.env['payment.transaction'].search([('reference', '=', reference)])
+        if not tx or len(tx) > 1:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "received data with reference %(ref)s matching %(num_tx)d transaction(s)",
+                    ref=reference, num_tx=len(tx)
+                )
+            )
+        return tx
+
+    def _process_feedback_data(self, data):
+        if self.provider != 'alipay':
+            return super()._process_feedback_data(data)
+        if self.state in ['done', 'cancel']:
+            _logger.info(f"Alipay: trying to validate an already terminated tx (ref {self.reference}), "
+                         f"state: {self.state}, received data: {data}")
+            return
+        if float_compare(float(data.get('total_fee', '0.0')), (self.amount + self.fees), 2) != 0:
+            # mc_gross is amount + fees
+            _logger.error(f"Alipay: The paid amount ({data.get('total_fee', '0.0')}) does not match the "
+                          f"total + fees ({self.amount} + {self.fees}) for ref {self.reference}")
```

Comment:
```suggestion
```
`ValidationError` already logs a warning. Given that the controller already logs the entire data, I guess that this log can be removed.